clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
         getLoc(Loc), load, LV.getVectorIdx()));
   }
 
+  if (LV.isExtVectorElt()) {
+    return buildLoadOfExtVectorElementLValue(LV);
+  }
+
   llvm_unreachable("NYI");
 }
 
+int64_t CIRGenFunction::getAccessedFieldNo(unsigned int idx,
+                                           const mlir::Attribute elts) {
+  assert(elts.isa<mlir::ArrayAttr>());
+
+  auto elt = elts.cast<mlir::ArrayAttr>()[idx];
+  assert(elt.isa<mlir::IntegerAttr>());
+  return elt.cast<mlir::IntegerAttr>().getInt();
+}
+
+// If this is a reference to a subset of the elements of a vector, create an
+// appropriate shufflevector.
+RValue CIRGenFunction::buildLoadOfExtVectorElementLValue(LValue LV) {
+  mlir::Location loc = LV.getExtVectorPointer().getLoc();
+  mlir::Value Vec = builder.createLoad(loc, LV.getExtVectorAddress());
+
+  // HLSL allows treating scalars as one-element vectors. Converting the scalar
+  // IR value to a vector here allows the rest of codegen to behave as normal.
+  if (getLangOpts().HLSL && !Vec.getType().isa<mlir::cir::VectorType>()) {
+    auto DstTy =
```

Comment:
Paths that are not tested should be `llvm_unreachable("NYI");`. Unless you are adding HLSL tests, please remove the code within this if statement in favor of `llvm_unreachable("NYI");`. Please do that for all other places where you didn't add tests to cover those paths.